internal/engine/upper.go

```diff
@@ -2,24 +2,63 @@ package engine
 
 import (
 	"context"
+	"errors"
+
 	"github.com/windmilleng/tilt/internal/logger"
 	"github.com/windmilleng/tilt/internal/model"
+	"github.com/windmilleng/tilt/internal/watch"
+
 	"io"
 )
 
-func UpService(ctx context.Context, service model.Service, watch bool, stdout io.Writer, stderr io.Writer) error {
-	bad, err := NewLocalBuildAndDeployer()
+type Upper struct {
+	b            BuildAndDeployer
+	watcherMaker func() (watch.Notify, error)
+}
+
+func NewUpper() (Upper, error) {
+	b, err := NewLocalBuildAndDeployer()
+	if err != nil {
+		return Upper{}, err
+	}
+	watcherMaker := func() (watch.Notify, error) {
+		return watch.NewWatcher()
+	}
+	return Upper{b, watcherMaker}, nil
+}
+
+func (u Upper) Up(ctx context.Context, service model.Service, watchMounts bool, stdout io.Writer, stderr io.Writer) error {
+	buildToken, err := u.b.BuildAndDeploy(ctx, service, nil)
 	if err != nil {
 		return err
 	}
-	buildToken, err := bad.BuildAndDeploy(ctx, service, nil)
-	if watch {
+
+	if watchMounts {
+		watcher, err := u.watcherMaker()
+		if err != nil {
+			return err
+		}
+
+		if len(service.Mounts) == 0 {
+			return errors.New("service has 0 repos - nothing to watch")
+		}
+
+		for _, mount := range service.Mounts {
+			watcher.Add(mount.Repo.LocalPath)
+		}
+
 		for {
-			// TODO(matt) actually wait for a file to change instead of building on a loop
-			logger.Get(ctx).Verbose("building and deploying")
-			buildToken, err = bad.BuildAndDeploy(ctx, service, buildToken)
-			if err != nil {
+			// TODO(matt) honor .gitignore / .dockerignore
+			// TODO(matt) buffer events a bit so that we're not triggering 10 builds when you change branches
+			select {
+			case err := <-watcher.Errors():
 				return err
+			case <-watcher.Events():
+				logger.Get(ctx).Verbose("file changed, rebuilding %v", service.Name)
+				buildToken, err = u.b.BuildAndDeploy(ctx, service, buildToken)
```

Comment:
if you're auto-rebuilding every time, it seems like...builds will fail often? and we'll want to build them next time?

Reply:
whoops, good catch